snuba/query/processors/uniq_in_select_and_having.py

```diff
@@ -0,0 +1,92 @@
+from snuba.state import get_config
+
+
+class MismatchedAggregationException(InvalidQueryException):
+    pass
+
+
+class _FunctionNameFindingVisitor(NoopVisitor):
+    def __init__(self, function_name_to_find: str):
+        self.function_name_to_find = function_name_to_find
+        self.found_functions: List[FunctionCall] = []
+
+    def visit_function_call(self, exp: FunctionCall) -> None:
+        for param in exp.parameters:
+            param.accept(self)
+        if exp.function_name == self.function_name_to_find:
+            self.found_functions.append(exp)
+        return None
+
+
+def _expressions_equal_ignore_aliases(exp1: Expression, exp2: Expression) -> bool:
+    if type(exp1) != type(exp2):
+        return False
+    fields_equal = True
+    for field in fields(exp1):
+        if field.name == "alias":
+            continue
+        field_val1 = getattr(exp1, field.name)
+        field_val2 = getattr(exp2, field.name)
+        if isinstance(field_val1, Expression):
+            fields_equal &= _expressions_equal_ignore_aliases(field_val1, field_val2)
+        else:
+            fields_equal &= field_val1 == field_val2
+        if not fields_equal:
+            return False
+    return fields_equal
+
+
+class _ExpressionOrAliasMatcher(NoopVisitor):
+    def __init__(self, expressions_to_match: Sequence[FunctionCall]):
+        self.expressions_to_match = expressions_to_match
+        self.found_expressions = [False] * len(expressions_to_match)
+
+    def visit_function_call(self, exp: FunctionCall) -> None:
+        for param in exp.parameters:
+            param.accept(self)
+        for i, exp_to_match in enumerate(self.expressions_to_match):
+            if (
+                exp_to_match.function_name == exp.function_name
+                and _expressions_equal_ignore_aliases(exp_to_match, exp)
+            ):
+                self.found_expressions[i] = True
+
+
+class UniqInSelectAndHavingProcessor(QueryProcessor):
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        having_clause = query.get_having()
+        if not having_clause:
+            return None
+        selected_columns = query.get_selected_columns()
+        uniq_finder = _FunctionNameFindingVisitor("uniq")
+        having_clause.accept(uniq_finder)
+        if uniq_finder.found_functions:
+            matcher = _ExpressionOrAliasMatcher(uniq_finder.found_functions)
+            for col in selected_columns:
+                col.expression.accept(matcher)
+            if not all(matcher.found_expressions):
+                should_throw = get_config("throw_on_uniq_select_and_having", False)
+                error = MismatchedAggregationException(
+                    f"Aggregation is in HAVING clause but not SELECT: {uniq_finder.found_functions}"
+                )
+                if should_throw:
+                    raise error
+                else:
+                    logging.warning(str(error))
```

Comment:
If you log the warning this way, every message with a different expression will cause a new group. Which can be a lot and also unreadable.

Please:
- use the `extra` field to provide metadata for your error, you can pass a dictionary.
- set `exc_info` = True so you have the stack trace and function call parameters in sentry
- Only write the error message as log message. Do not serialize the parts of the AST inside the message